### PR TITLE
Feature: Skip blank tiles when creating XYZ tiles

### PR DESCRIPTION
**Pull request by Soevndahl**
## Description

Fix issue #35174 Skipping blank tiles when creating XYZ tiles
When generating tiles, if Skip Empty Tiles is checked, the tiles will not be saved in the directory or in the mbtiles file.
The algorithm is still slow for large areas but this saves space because all empty tiles are not saved.

### CODE REVIEW

**Comment by m-kuhn** [comment]
@alexbruy what do you think about this?

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 1]
```diff
@@ -111,6 +111,7 @@ void QgsXyzTilesBaseAlgorithm::createCommonParameters()
   addParameter( new QgsProcessingParameterNumber( QStringLiteral( "DPI" ), QObject::tr( "DPI" ), QgsProcessingParameterNumber::Integer, 96, false, 48, 600 ) );
   addParameter( new QgsProcessingParameterColor( QStringLiteral( "BACKGROUND_COLOR" ), QObject::tr( "Background color" ), QColor( Qt::transparent ), true, true ) );
   addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "ANTIALIAS" ), QObject::tr( "Enable antialiasing" ), true ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "Skip Empty Tiles" ), QObject::tr( "Skip empty tiles" ), true ) );
```
Comment (alexbruy) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:114]: We need to stick with parameter naming conventions. Also let's make it optional to keep existing script/models working.

```suggestion
  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "SKIP_EMPTY_TILES" ), QObject::tr( "Skip empty tiles" ), true, true ) );
```

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 2]
```diff
@@ -153,13 +154,23 @@ bool QgsXyzTilesBaseAlgorithm::prepareAlgorithm( const QVariantMap &parameters,
   mDpi = parameterAsInt( parameters, QStringLiteral( "DPI" ), context );
   mBackgroundColor = parameterAsColor( parameters, QStringLiteral( "BACKGROUND_COLOR" ), context );
   mAntialias = parameterAsBool( parameters, QStringLiteral( "ANTIALIAS" ), context );
+  mSkipEmptyTiles = parameterAsBool( parameters, QStringLiteral( "Skip Empty Tiles" ), context );
```
Comment (alexbruy) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:157]:
```suggestion
  mSkipEmptyTiles = parameterAsBool( parameters, QStringLiteral( "SKIP_EMPTY_TILES" ), context );
```

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 3]
```diff
@@ -376,13 +387,28 @@ void QgsXyzTilesDirectoryAlgorithm::processMetaTile( QgsMapRendererSequentialJob
...
+    if (mSkipEmptyTiles == true)
+    {
+      if ( tileImage == mBackgroundImage)
+      {
+        ++it;
+        continue;
+      }
+    }
+
+
```
Comment (alexbruy) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:400]: Remove extra line

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 4]
```diff
@@ -376,13 +387,28 @@ void QgsXyzTilesDirectoryAlgorithm::processMetaTile( QgsMapRendererSequentialJob
...
     tileDir.mkpath( tileDir.absolutePath() );
     int y = tile.y;
     if ( mTms )
     {
       y = tile2tms( y, tile.z );
     }
+
+
+
+
```
Comment (alexbruy) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:411]: Remove extra lines

**Review by alexbruy — CHANGES_REQUESTED** [review]
May I also ask you to squash commits?

As for the comparison of `QImages`, Qt documentation states

> The comparison can be slow, unless there is some obvious difference (e.g. different size or format), in which case the function will return quickly.

So enabling this may make the algorithm much more slower. Maybe consider some other ways to detect empty tiles. In case if we will stick with this approach, I would suggest mentioning expected slowdown in the algorithm help.

**Comment by agiudiceandrea** [comment]
> enabling this may make the algorithm much more slower

It seems the current C++ version (in 3,34) of the algorithm for MBTiles is already much more slower than the Python one (in 3.28) according to https://github.com/qgis/QGIS/issues/55489.

**Comment by agiudiceandrea** [comment]
@Soevndahl , please have a look at the requested changes.

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 5]
```diff
@@ -111,7 +111,7 @@ void QgsXyzTilesBaseAlgorithm::createCommonParameters()
   addParameter( new QgsProcessingParameterNumber( QStringLiteral( "DPI" ), QObject::tr( "DPI" ), QgsProcessingParameterNumber::Integer, 96, false, 48, 600 ) );
   addParameter( new QgsProcessingParameterColor( QStringLiteral( "BACKGROUND_COLOR" ), QObject::tr( "Background color" ), QColor( Qt::transparent ), true, true ) );
   addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "ANTIALIAS" ), QObject::tr( "Enable antialiasing" ), true ) );
-  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "Skip Empty Tiles" ), QObject::tr( "Skip empty tiles" ), true ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "Skip_Empty_Tiles" ), QObject::tr( "Skip empty tiles" ), true, true ) );
```
Comment (m-kuhn) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:114]:
```suggestion
  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "SKIP_EMPTY_TILES" ), QObject::tr( "Skip empty tiles" ), true, true ) );
```

let's be consistent with capitalization

File: src/analysis/processing/qgsalgorithmxyztiles.cpp [hunk 6]
```diff
@@ -154,7 +154,7 @@ bool QgsXyzTilesBaseAlgorithm::prepareAlgorithm( const QVariantMap &parameters,
   mDpi = parameterAsInt( parameters, QStringLiteral( "DPI" ), context );
   mBackgroundColor = parameterAsColor( parameters, QStringLiteral( "BACKGROUND_COLOR" ), context );
   mAntialias = parameterAsBool( parameters, QStringLiteral( "ANTIALIAS" ), context );
-  mSkipEmptyTiles = parameterAsBool( parameters, QStringLiteral( "Skip Empty Tiles" ), context );
+  mSkipEmptyTiles = parameterAsBool( parameters, QStringLiteral( "Skip_Empty_Tiles" ), context );
```
Comment (m-kuhn) [inline src/analysis/processing/qgsalgorithmxyztiles.cpp:157]:
```suggestion
  mSkipEmptyTiles = parameterAsBool( parameters, QStringLiteral( "SKIP_EMPTY_TILES" ), context );
```

**Comment by agiudiceandrea** [comment]
@Soevndahl , please have a look at the requested changes.

**Pull request closed without merging**